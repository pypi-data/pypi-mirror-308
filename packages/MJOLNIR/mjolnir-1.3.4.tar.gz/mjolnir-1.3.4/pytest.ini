[pytest]
python_files = test/*.py
markers =
    unit