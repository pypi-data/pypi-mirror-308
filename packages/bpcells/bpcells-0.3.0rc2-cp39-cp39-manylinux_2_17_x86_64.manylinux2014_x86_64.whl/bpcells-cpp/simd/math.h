// Copyright 2023 BPCells contributors
// 
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#pragma once


namespace BPCells::simd {

void log1p(float *inout, size_t n);
void log1p(double *inout, size_t n);
void log1p_downcast(double *inout, size_t n);

void expm1(float *inout, size_t n);
void expm1(double *inout, size_t n);
void expm1_downcast(double *inout, size_t n);

void square(float *inout, size_t n);
void square(double *inout, size_t n);
void square_downcast(double *inout, size_t n);

// Return the max value of `in`
uint32_t max(const uint32_t *in, size_t n);

// Write inout[i] = inout[i] + a[i]
void add(uint32_t *inout, const uint32_t *a, size_t n);

// Write inout[i] = inout[i] + a
void add(uint32_t *inout, const int32_t a, size_t n);

// Write inout[i] = inout[i] - a[i]
void sub(uint32_t *inout, const uint32_t *a, size_t n);

}