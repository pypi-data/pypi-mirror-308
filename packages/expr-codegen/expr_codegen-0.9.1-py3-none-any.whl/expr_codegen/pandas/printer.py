from sympy import Basic, Function, StrPrinter
from sympy.printing.precedence import precedence, PRECEDENCE


# TODO: 如有新添加函数，但表达式有变更才需要在此补充对应的打印代码，否则可以省略

class PandasStrPrinter(StrPrinter):
    def _print(self, expr, **kwargs) -> str:
        """Internal dispatcher

        Tries the following concepts to print an expression:
            1. Let the object print itself if it knows how.
            2. Take the best fitting method defined in the printer.
            3. As fall-back use the emptyPrinter method for the printer.
        """
        self._print_level += 1
        try:
            # If the printer defines a name for a printing method
            # (Printer.printmethod) and the object knows for itself how it
            # should be printed, use that method.
            if self.printmethod and hasattr(expr, self.printmethod):
                if not (isinstance(expr, type) and issubclass(expr, Basic)):
                    return getattr(expr, self.printmethod)(self, **kwargs)

            # See if the class of expr is known, or if one of its super
            # classes is known, and use that print function
            # Exception: ignore the subclasses of Undefined, so that, e.g.,
            # Function('gamma') does not get dispatched to _print_gamma
            classes = type(expr).__mro__
            # if AppliedUndef in classes:
            #     classes = classes[classes.index(AppliedUndef):]
            # if UndefinedFunction in classes:
            #     classes = classes[classes.index(UndefinedFunction):]
            # Another exception: if someone subclasses a known function, e.g.,
            # gamma, and changes the name, then ignore _print_gamma
            if Function in classes:
                i = classes.index(Function)
                classes = tuple(c for c in classes[:i] if \
                                c.__name__ == classes[0].__name__ or \
                                c.__name__.endswith("Base")) + classes[i:]
            # print('=' * 60)
            for cls in classes:
                printmethodname = '_print_' + cls.__name__
                # print(printmethodname, expr)
                printmethod = getattr(self, printmethodname, None)
                if printmethod is not None:
                    return printmethod(expr, **kwargs)
            # Unknown object, fall back to the emptyPrinter.
            return self.emptyPrinter(expr)
        finally:
            self._print_level -= 1

    def _print_Symbol(self, expr):
        return f"df[{expr.name}]"

    def _print_Equality(self, expr):
        PREC = precedence(expr)
        return "%s==%s" % (self.parenthesize(expr.args[0], PREC), self.parenthesize(expr.args[1], PREC))

    def _print_Or(self, expr):
        PREC = PRECEDENCE["Mul"]
        return " | ".join(self.parenthesize(arg, PREC) for arg in expr.args)

    def _print_Xor(self, expr):
        PREC = PRECEDENCE["Mul"]
        return " ^ ".join(self.parenthesize(arg, PREC) for arg in expr.args)

    def _print_And(self, expr):
        PREC = PRECEDENCE["Mul"]
        return " & ".join(self.parenthesize(arg, PREC) for arg in expr.args)

    def _print_Not(self, expr):
        PREC = PRECEDENCE["Mul"]
        return "~%s" % self.parenthesize(expr.args[0], PREC)

    def _print_if_else(self, expr):
        return "np.where(%s, %s, %s)" % (self._print(expr.args[0]), self._print(expr.args[1]), self._print(expr.args[2]))

    def _print_ts_mean(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).mean()" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_std_dev(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).std(ddof=0)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_arg_max(self, expr):
        return "bn.move_argmax(%s, %s)" % (self._print(expr.args[0]), self._print(expr.args[1]))

    def _print_ts_arg_min(self, expr):
        return "bn.move_argmin(%s, %s)" % (self._print(expr.args[0]), self._print(expr.args[1]))

    def _print_ts_product(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).apply(np.product, raw=True)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_max(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).max()" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_min(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).min()" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_delta(self, expr):
        PREC = precedence(expr)
        return "%s.diff(%s)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_delay(self, expr):
        PREC = precedence(expr)
        return "%s.shift(%s)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_corr(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).corr(%s, ddof=0)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[2]), self._print(expr.args[1]))

    def _print_ts_covariance(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).cov(%s, ddof=0)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[2]), self._print(expr.args[1]))

    def _print_ts_rank(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).rank(pct=True)" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_sum(self, expr):
        PREC = precedence(expr)
        return "%s.rolling(%s).sum()" % (self.parenthesize(expr.args[0], PREC), self._print(expr.args[1]))

    def _print_ts_decay_linear(self, expr):
        return "ta.WMA(%s, %s)" % (self._print(expr.args[0]), self._print(expr.args[1]))

    def _print_cs_rank(self, expr):
        PREC = precedence(expr)
        return "%s.rank(pct=True)" % self.parenthesize(expr.args[0], PREC)

    def _print_cs_scale(self, expr):
        return "scale(%s)" % self._print(expr.args[0])

    def _print_log(self, expr):
        return "np.log(%s)" % self._print(expr.args[0])

    def _print_abs_(self, expr):
        PREC = precedence(expr)
        if expr.args[0].is_Number:
            return "np.abs(%s)" % expr.args[0]
        else:
            return "%s.abs()" % self.parenthesize(expr.args[0], PREC)

    def _print_max_(self, expr):
        return "np.maximum(%s, %s)" % (self._print(expr.args[0]), self._print(expr.args[1]))

    def _print_min_(self, expr):
        return "np.minimum(%s, %s)" % (self._print(expr.args[0]), self._print(expr.args[1]))

    def _print_sign(self, expr):
        return "np.sign(%s)" % self._print(expr.args[0])

    def _print_gp_rank(self, expr):
        PREC = precedence(expr)
        return "%s.rank(pct=True)" % self.parenthesize(expr.args[0], PREC)

    def _print_gp_neutralize(self, expr):
        return "neutralize(%s)" % self._print(expr.args[1])
