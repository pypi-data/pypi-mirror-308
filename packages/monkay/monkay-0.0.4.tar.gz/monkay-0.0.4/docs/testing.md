# Testing


## Temporary overwrites

For tests, but not limited to, Monkay provides three methods returning a contextmanager which provides threadsafe a temporary overwrite:

- `with_settings(settings)`
- `with_extensions(extensions_dict, *, apply_extensions=False)`
- `with_instance(instance, * apply_extensions=False,use_extensions_overwrite=True)`


## Check lazy imports

Monkay provides the debug method `find_missing(*, all_var=None, search_pathes=None, ignore_deprecated_import_errors=False, require_search_path_all_var=True)`.
It is quite expensive so it should be only called for debugging, testing and in error cases.
It returns a dictionary containing items which had issues, e.g. imports failed or not in `__all__` variable.

When providing `search_pathes` (module pathes as string), all exports are checked if they are in the value set of Monkey.

When providing `__all__` as `all_var`, it is checked for all imports.

Returned is a dictionary in the format:

- key: import name or import path
- value: set with errors

Errors:

- `all_var`: key is not in the provided `__all__` variable
- `import`: key had an ImportError
- `search_path_extra`: key (here a path) is not included in lazy imports.
- `search_path_import`: import of key (here the search path) failed
- `search_path_all_var`: module imported as search path had no `__all__`. This error can be disabled with `require_search_path_all_var=False`

### Ignore import errors when lazy import is deprecated

The parameter `ignore_deprecated_import_errors=True` silences errors happening when an lazy import which was marked as deprecated failed.

### Example

Using Monkay for tests is confortable and easy:

``` python

import edgy

def test_edgy_lazy_imports():
    assert not edgy.monkay.find_missing(all_var=edgy.__all__, search_pathes=["edgy.core.files", "edgy.core.db.fields", "edgy.core.connection"])

```

That was the test. Now we know that no lazy import is broken.
