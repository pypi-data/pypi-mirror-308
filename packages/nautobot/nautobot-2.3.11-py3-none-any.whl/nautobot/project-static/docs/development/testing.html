
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>Redirecting...</title>
    <link rel="canonical" href="core/testing.html">
    <meta name="robots" content="noindex">
    <script>var anchor=window.location.hash.substr(1);location.href="core/testing.html"+(anchor?"#"+anchor:"")</script>
    <meta http-equiv="refresh" content="0; url=core/testing.html">
</head>
<body>
Redirecting...
</body>
</html>
