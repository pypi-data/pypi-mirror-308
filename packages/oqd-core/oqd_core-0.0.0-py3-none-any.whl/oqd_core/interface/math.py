# Copyright 2024 Open Quantum Design

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import ast
import numpy as np
from typing import Union, Literal, Annotated, Any
from pydantic import AfterValidator, BeforeValidator

from oqd_compiler_infrastructure import TypeReflectBaseModel, ConversionRule

########################################################################################

__all__ = [
    "MathExpr",
    "MathTerminal",
    "MathStr",
    "MathNum",
    "MathVar",
    "MathImag",
    "MathUnaryOp",
    "MathFunc",
    "MathBinaryOp",
    "MathAdd",
    "MathSub",
    "MathMul",
    "MathDiv",
    "MathPow",
    "MathExprSubtypes",
]

########################################################################################


class MathExpr(TypeReflectBaseModel):
    """
    Class representing the abstract syntax tree (AST) for a mathematical expression
    """

    @classmethod
    def cast(cls, value: Any):
        if isinstance(value, MathExpr):
            return value
        if isinstance(value, (int, float)):
            value = MathNum(value=value)
            return value
        if isinstance(value, (complex, np.complex128)):
            value = MathNum(value=value.real) + MathImag() * value.imag
            return value
        if isinstance(value, str):
            raise TypeError(
                "Tried to cast a string to MathExpr. "
                + f'Wrap your string ("{value}") with MathStr(string="{value}").'
            )
        raise TypeError

    def __neg__(self):
        return MathMul(expr1=MathNum(value=-1), expr2=self)

    def __pos__(self):
        return self

    def __add__(self, other):
        return MathAdd(expr1=self, expr2=other)

    def __sub__(self, other):
        return MathSub(expr1=self, expr2=other)

    def __mul__(self, other):
        try:
            return MathMul(expr1=self, expr2=other)
        except:
            return other * self

    def __truediv__(self, other):
        return MathDiv(expr1=self, expr2=other)

    def __pow__(self, other):
        return MathPow(expr1=self, expr2=other)

    def __radd__(self, other):
        other = MathExpr.cast(other)
        return other + self

    def __rsub__(self, other):
        other = MathExpr.cast(other)
        return other - self

    def __rmul__(self, other):
        other = MathExpr.cast(other)
        return other * self

    def __rpow__(self, other):
        other = MathExpr.cast(other)
        return other**self

    def __rtruediv__(self, other):
        other = MathExpr.cast(other)
        return other / self

    pass


########################################################################################


def is_varname(value: str) -> str:
    if not value.isidentifier():
        raise ValueError
    return value


VarName = Annotated[str, AfterValidator(is_varname)]
CastMathExpr = Annotated[MathExpr, BeforeValidator(MathExpr.cast)]
Functions = Literal["sin", "cos", "tan", "exp", "log", "sinh", "cosh", "tanh"]


########################################################################################


class AST_to_MathExpr(ConversionRule):
    def generic_map(self, model: Any, operands):
        raise TypeError

    def map_Module(self, model: ast.Module, operands):
        if len(model.body) == 1:
            return self(model.body[0])
        raise TypeError

    def map_Expr(self, model: ast.Expr, operands):
        return self(model.value)

    def map_Constant(self, model: ast.Constant, operands):
        return MathExpr.cast(model.value)

    def map_Name(self, model: ast.Name, operands):
        return MathVar(name=model.id)

    def map_BinOp(self, model: ast.BinOp, operands):
        if isinstance(model.op, ast.Add):
            return MathAdd(expr1=self(model.left), expr2=self(model.right))
        if isinstance(model.op, ast.Sub):
            return MathSub(expr1=self(model.left), expr2=self(model.right))
        if isinstance(model.op, ast.Mult):
            return MathMul(expr1=self(model.left), expr2=self(model.right))
        if isinstance(model.op, ast.Div):
            return MathDiv(expr1=self(model.left), expr2=self(model.right))
        if isinstance(model.op, ast.Pow):
            return MathPow(expr1=self(model.left), expr2=self(model.right))
        raise TypeError

    def map_UnaryOp(self, model: ast.UnaryOp, operands):
        if isinstance(model.op, ast.USub):
            return -self(model.operand)
        if isinstance(model.op, ast.UAdd):
            return self(model.operand)
        raise TypeError

    def map_Call(self, model: ast.Call, operands):
        if len(model.args) == 1:
            return MathFunc(func=model.func.id, expr=self(model.args[0]))
        raise TypeError


def MathStr(*, string):
    return AST_to_MathExpr()(ast.parse(string))


########################################################################################


class MathTerminal(MathExpr):
    """
    Class representing a terminal in the [`MathExpr`][oqd_core.interface.math.MathExpr] abstract syntax tree (AST)
    """

    pass


class MathVar(MathTerminal):
    """
    Class representing a variable in a [`MathExpr`][oqd_core.interface.math.MathExpr]

    Examples:
        >>> MathVar("t")

    """

    name: VarName


class MathNum(MathTerminal):
    """
    Class representing a number in a [`MathExpr`][oqd_core.interface.math.MathExpr]
    """

    value: Union[int, float]


class MathImag(MathTerminal):
    """
    Class representing the imaginary unit in a [`MathExpr`][oqd_core.interface.math.MathExpr] abstract syntax tree (AST)
    """

    pass


class MathUnaryOp(MathExpr):
    """
    Class representing a unary operations on a [`MathExpr`][oqd_core.interface.math.MathExpr] abstract syntax tree (AST)
    """

    pass


class MathFunc(MathUnaryOp):
    """
    Class representing a named function applied to a [`MathExpr`][oqd_core.interface.math.MathExpr] abstract syntax tree (AST)

    Attributes:
        func (Literal["sin", "cos", "tan", "exp", "log", "sinh", "cosh", "tanh"]): Named function to apply
        expr (MathExpr): Argument of the named function
    """

    func: Functions
    expr: CastMathExpr


class MathBinaryOp(MathExpr):
    """
    Class representing binary operations on [`MathExprs`][oqd_core.interface.math.MathExpr] abstract syntax tree (AST)
    """

    pass


class MathAdd(MathBinaryOp):
    """
    Class representing the addition of [`MathExprs`][oqd_core.interface.analog.operator.Operator]

    Attributes:
        expr1 (MathExpr): Left hand side [`MathExpr`][oqd_core.interface.analog.operator.Operator]
        expr2 (MathExpr): Right hand side [`MathExpr`][oqd_core.interface.analog.operator.Operator]
    """

    expr1: CastMathExpr
    expr2: CastMathExpr


class MathSub(MathBinaryOp):
    """
    Class representing the subtraction of [`MathExprs`][oqd_core.interface.math.MathExpr]

    Attributes:
        expr1 (MathExpr): Left hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
        expr2 (MathExpr): Right hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
    """

    expr1: CastMathExpr
    expr2: CastMathExpr


class MathMul(MathBinaryOp):
    """
    Class representing the multiplication of [`MathExprs`][oqd_core.interface.math.MathExpr]

    Attributes:
        expr1 (MathExpr): Left hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
        expr2 (MathExpr): Right hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
    """

    expr1: CastMathExpr
    expr2: CastMathExpr


class MathDiv(MathBinaryOp):
    """
    Class representing the division of [`MathExprs`][oqd_core.interface.math.MathExpr]

    Attributes:
        expr1 (MathExpr): Left hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
        expr2 (MathExpr): Right hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
    """

    expr1: CastMathExpr
    expr2: CastMathExpr


class MathPow(MathBinaryOp):
    """
    Class representing the exponentiation of [`MathExprs`][oqd_core.interface.math.MathExpr]

    Attributes:
        expr1 (MathExpr): Left hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
        expr2 (MathExpr): Right hand side [`MathExpr`][oqd_core.interface.math.MathExpr]
    """

    expr1: CastMathExpr
    expr2: CastMathExpr


########################################################################################

MathExprSubtypes = Union[
    MathNum,
    MathVar,
    MathImag,
    MathFunc,
    MathAdd,
    MathSub,
    MathMul,
    MathDiv,
    MathPow,
]
"""
Alias for the union of concrete MathExpr subtypes
"""
