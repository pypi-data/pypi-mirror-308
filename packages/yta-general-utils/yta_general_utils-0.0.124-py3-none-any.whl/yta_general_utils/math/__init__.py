import numpy as np


class Math:
    """
    Class to simplify and encapsulate functionality related
    to math that could be more complex than the contained 
    in the basic python math module.
    """
    @staticmethod
    def sigmoid(value):
        """
        TODO: Write doc about this
        """
        return 1.0 / (1 + np.exp(-value))