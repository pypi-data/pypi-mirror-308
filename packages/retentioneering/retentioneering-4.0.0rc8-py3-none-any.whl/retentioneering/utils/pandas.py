# type: ignore

import random

import pandas as pd


def shuffle_df(df: pd.DataFrame, inplace=True):
    df = df if inplace else df.copy()
    index = [i for i in range(df.shape[0])]
    random.shuffle(index)
    df.set_index([index]).sort_index()
    return df


def get_merged_col(df: pd.DataFrame, colname: str, suffix: str):
    return df[colname] if colname in df else df[f"{colname}{suffix}"]


__all__ = ["shuffle_df", "get_merged_col"]
