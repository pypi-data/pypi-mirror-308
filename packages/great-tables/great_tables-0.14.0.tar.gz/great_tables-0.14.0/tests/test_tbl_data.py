import math
import pandas as pd
import polars as pl
import polars.testing
import pytest
from great_tables._tbl_data import (
    DataFrameLike,
    SeriesLike,
    _get_cell,
    _get_column_dtype,
    _set_cell,
    _validate_selector_list,
    cast_frame_to_string,
    create_empty_frame,
    eval_select,
    get_column_names,
    group_splits,
    is_series,
    reorder,
    to_frame,
    validate_frame,
)

params_frames = [pytest.param(pd.DataFrame, id="pandas"), pytest.param(pl.DataFrame, id="polars")]
params_series = [pytest.param(pd.Series, id="pandas"), pytest.param(pl.Series, id="polars")]


@pytest.fixture(params=params_frames, scope="function")
def df(request) -> pd.DataFrame:
    return request.param({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})


@pytest.fixture(params=params_series, scope="function")
def ser(request) -> SeriesLike:
    return request.param([1.0, 2.0, None])


def assert_frame_equal(src, target):
    if isinstance(src, pd.DataFrame):
        pd.testing.assert_frame_equal(src, target)
    elif isinstance(src, pl.DataFrame):
        pl.testing.assert_frame_equal(src, target)
    else:
        raise NotImplementedError(f"Unsupported data type: {type(src)}")


def test_get_column_names(df: DataFrameLike):
    expected = ["col1", "col2", "col3"]
    assert get_column_names(df) == expected


def test_get_column_dtypes(df: DataFrameLike):
    assert _get_column_dtype(df, "col1") == df["col1"].dtype


def test_get_cell(df: DataFrameLike):
    assert _get_cell(df, 1, "col2") == "b"


def test_set_cell(df: DataFrameLike):
    expected = df.__class__({"col1": [1, 2, 3], "col2": ["a", "x", "c"], "col3": [4.0, 5.0, 6.0]})
    _set_cell(df, 1, "col2", "x")
    assert_frame_equal(df, expected)


def test_reorder(df: DataFrameLike):
    res = reorder(df, [0, 2], ["col2"])
    dst = df.__class__({"col2": ["a", "c"]})

    if isinstance(dst, pd.DataFrame):
        dst.index = pd.Index([0, 2])

    assert_frame_equal(res, dst)


@pytest.mark.parametrize("expr", [["col2", "col1"], [1, 0], ["col2", 0], [1, "col1"]])
def test_eval_select_with_list(df: DataFrameLike, expr):
    sel = eval_select(df, expr)
    assert sel == [("col2", 1), ("col1", 0)]


@pytest.mark.parametrize(
    "expr",
    [
        pl.selectors.exclude("col3"),
        pl.selectors.starts_with("col1") | pl.selectors.starts_with("col2"),
        [pl.col("col1"), pl.col("col2")],
        [pl.col("col1"), pl.selectors.by_name("col2")],
        pl.col("col1", "col2"),
        pl.all().exclude("col3"),
    ],
)
def test_eval_select_with_list_pl_selector(expr):
    df = pl.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    sel = eval_select(df, expr)
    assert sel == [("col1", 0), ("col2", 1)]


@pytest.mark.parametrize("expr", [["col2", 1.2]])
def test_eval_select_pandas_raises1(expr):
    df = pd.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    with pytest.raises(TypeError) as exc_info:
        eval_select(df, expr)

    assert "Only int and str are supported." in str(exc_info.value.args[0])


@pytest.mark.parametrize("expr", [3.45, {"col2"}, ("col2",)])
def test_eval_select_pandas_raises2(expr):
    df = pd.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    with pytest.raises(NotImplementedError) as exc_info:
        eval_select(df, expr)

    assert "Unsupported selection expr: " in str(exc_info.value.args[0])


@pytest.mark.parametrize("expr", [3.45, {6}, (7.8,)])
def test_eval_select_polars_raises(expr):
    df = pl.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    with pytest.raises(TypeError) as exc_info:
        eval_select(df, expr)

    assert "Unsupported selection expr type:" in str(exc_info.value.args[0])


def test_eval_selector_polars_list_raises():
    expr = ["col1", 1.2]
    df = pl.DataFrame({"col1": [], "col2": [], "col3": []})
    with pytest.raises(TypeError) as exc_info:
        eval_select(df, expr)

    assert "entry 1 is type: <class 'float'>" in str(exc_info.value.args[0])


@pytest.mark.parametrize("Frame", [pd.DataFrame, pl.DataFrame])
def test_group_splits_pd(Frame):
    df = Frame({"g": ["b", "a", "b", "c"]})

    splits = group_splits(df, "g")
    assert set(splits.keys()) == {"a", "b", "c"}
    assert splits["b"] == [0, 2]
    assert splits["a"] == [1]
    assert splits["c"] == [3]


def test_group_splits_pd_na():
    df = pd.DataFrame({"g": ["b", "a", None]})

    splits = group_splits(df, "g")
    assert len(splits.keys()) == 3
    nan_key = [k for k in splits if isinstance(k, float) and math.isnan(k)][0]

    assert splits[nan_key] == [2]


def test_group_splits_pl_na():
    df = pl.DataFrame({"g": ["b", "a", None]})

    splits = group_splits(df, "g")
    assert set(splits.keys()) == {"b", "a", None}
    assert splits[None] == [2]


def test_validate_selector_list_strict_raises():
    with pytest.raises(TypeError) as exc_info:
        _validate_selector_list([pl.col("a")])

    msg = "entry 0 is a polars Expr, which is only supported for polars versions >= 0.20.30."
    assert msg in str(exc_info.value.args[0])


def test_create_empty_frame(df: DataFrameLike):
    res = create_empty_frame(df)
    col = [None] * 3

    if isinstance(res, pd.DataFrame):
        dst = pd.DataFrame({"col1": col, "col2": col, "col3": col}, dtype="string")
    else:
        dst = pl.DataFrame({"col1": col, "col2": col, "col3": col}).cast(pl.Utf8)

    assert_frame_equal(res, dst)


def test_validate_frame_dupe_cols():
    df = pd.DataFrame([[1, 2, 3]], columns=["x", "x", "y"])

    with pytest.raises(ValueError) as exc_info:
        validate_frame(df)

    assert "Column names must be unique" in str(exc_info.value.args[0])


def test_validate_frame_multi_index():
    df = pd.DataFrame(
        [[1, 2, 3]], columns=pd.MultiIndex.from_tuples([("a", "x"), ("a", "y"), ("b", "x")])
    )

    with pytest.raises(ValueError) as exc_info:
        validate_frame(df)

    assert "MultiIndex columns are not supported" in str(exc_info.value.args[0])


def test_validate_frame_non_str_cols_warning(snapshot):
    df = pd.DataFrame({"x": [1], 55: [2], "y": [3], 99: [4]})

    with pytest.warns(UserWarning) as record:
        validate_frame(df)

    assert len(record) == 1
    assert snapshot == record[0].message.args[0]


def test_validate_frame_non_str_cols_result():
    df = pd.DataFrame({"x": [1], 55: [2], "y": [3], 99: [4]})

    with pytest.warns(UserWarning):
        res = validate_frame(df)

    assert list(res.columns) == ["x", "55", "y", "99"]


def test_to_frame(ser: SeriesLike):
    df = to_frame(ser, name="x")

    if isinstance(ser, pl.Series):
        assert_frame_equal(df, pl.DataFrame({"x": [1.0, 2.0, None]}))
    elif isinstance(ser, pd.Series):
        assert_frame_equal(df, pd.DataFrame({"x": [1.0, 2.0, None]}))
    else:
        raise AssertionError(f"Unexpected series type: {type(ser)}")


def test_is_series(ser: SeriesLike):
    assert is_series(ser)


def test_is_series_false():
    assert not is_series(1)


def test_cast_frame_to_string_polars_list_col():
    df = pl.DataFrame({"x": [[1, 2], [3]], "y": [1, None], "z": [{"a": 1}, {"a": 2}]})
    new_df = cast_frame_to_string(df)

    assert new_df["x"].dtype.is_(pl.String)
    assert new_df["y"].dtype.is_(pl.String)
    assert new_df["z"].dtype.is_(pl.String)
