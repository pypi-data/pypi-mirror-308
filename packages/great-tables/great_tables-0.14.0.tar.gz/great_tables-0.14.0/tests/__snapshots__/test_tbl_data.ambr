# serializer version: 1
# name: test_validate_frame_non_str_cols_warning
  '''
  pandas DataFrame contains non-string column names. Coercing to strings. Here are the first few non-string columns:
  
    * Position 1: 55
    * Position 3: 99
  '''
# ---
